### PR TITLE
Support musllinux binary wheels on x64 and x86

### PR DESCRIPTION
**Pull request by lidizheng**
PyPA's wheel standard has welcomed musl-libc based binary wheels. This PR adds support to build and test musllinux binary wheels for gRPC Python, in x64/x86.

AArch64 might need to wait for: 1. we have more accessible ARM compute resources; 2. Dockcross add "musllinux" (https://github.com/dockcross/dockcross) (or we build an image with cross-compiling CPython from source).

Fixes https://github.com/grpc/grpc/issues/27940 https://github.com/grpc/grpc/issues/26620 https://github.com/grpc/grpc/issues/25036
Related https://github.com/grpc/grpc/issues/27512 

Here are some implementation details about this PR:

- `mktemp` in Alpine works differently than Debian-based distros. It demands `XXXX` in the template of the temporary file/folder's name to always be suffix.
- We used to rename "grpcio-tools" to "grpc.tools", which is already deprecated (see [guide](https://grpc.io/docs/languages/python/quickstart/#generate-grpc-code)). The correct import name is "grpc_tools" (...why it isn't grpcio-tools at the first place). "grpc.tools" relies on an [operation](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/__init__.py#L2171) in the "grpcio" package. This PR corrected all "grpcio-tools" import to "grpc_tools".
- In the Alpine Python image, strangely, the wheel install complains about missing `libc++.so.6`. Technically, installing pip should include a compiler. This PR statically links the libc++.
- This PR extends the timeout of the Linux build_artifacts job from 180 to 240. Today, this job takes ~160 minutes to finish. With the added artifacts, it takes ~180 minutes, just crossed the timeout. The compilation time is optimized by @jtattermusch  before, the artifact building are running in parallel and using multiple cores. If we want to reduce the time, we  might need bigger machine, or split the linux build...

### CODE REVIEW

**Review by jtattermusch — CHANGES_REQUESTED** [review]
While it's nice that we could add musl linux support, I think the number of python wheels we build is really getting out of hand. The builds are already very slow and when building multilang artifacts, most of the time is spend building python wheels (which is both slow and there's a crazy number of different configurations).

Increasing the runtime of some of the PR jobs that are already horribly slow (e.g. build_artifacts) is counterproductive, has negative effect on everyone who's developing grpc and it kills all the gains we get in the currently ongoing effort to make the PR builds faster.

I think we need to stop and figure out how to make the python artifacts build much faster (which is the preferred solution) or at least how not to gate PRs on the build (but without introducing extra complexity and without risking breakages on master).  Until that happens, I'm strongly against adding anything to the build_artifacts job.

**Comment by gnossen** [comment]
Is parallelizing an option here?

**Comment by lidizheng** [comment]
We should be able to leverage the cloud resources we have. It should be possible to further parallelize the workload.

**Comment by allenporter** [comment]
> We should be able to leverage the cloud resources we have. It should be possible to further parallelize the workload.

One other option to consider is to dramatically cut the # of python versions supported for musl? Just supporting a few recent versions could help reduce the cost. Given alpine linux has an end of life support https://endoflife.date/alpine that currently includes only up to `v3.12` which includes [python 3.8.10](https://pkgs.alpinelinux.org/packages?name=python3&branch=v3.12) it seems nice to drop support for 3.6 and 3.7.

There is such a huge win to getting this rolled out, so thank you so much for pushing this forward.

**Comment by lidizheng** [comment]
@jtattermusch Do you think reducing 40% of musllinux wheels is enough for this change? We could keep the 180 minutes build time with this immediate change.

On the other hand, I'm exploring Cloud Build and GitHub Action. Yes, I recall go/grpc-ci-options, the doc didn't conclude which platform we should use. But for the artifact/package/distribution-test workflow, I think GitHub Action works better. It offers Linux/Win/Mac, as well as ARM support.

Can you give more direct suggestion about how should we push this PR forward?

**Comment by lidizheng** [comment]
@jtattermusch Also, are you open to split the artifact build job into multiple ones?

**Comment by jtattermusch** [comment]
@lidizheng I don't think there's a simple answer. I think we need to discuss offline (and brainstorm some ideas) and we possibly need a short doc to describe the path we want to take. There are too many possible gotchas to decide on an approach in a few comments on github. Also, any changes we make will likely have impact on our release process (which is already quite complicated so we should be simplifying things if possible).

**Comment by lidizheng** [comment]
@jtattermusch PTALA. The failing checks were introduced by recent Ruby issue. This PR has been up-to-date with recent infra changes.

File: tools/run_tests/artifacts/artifact_targets.py [hunk 1]
```diff
@@ -403,6 +419,16 @@ def targets():
         PythonArtifact('linux_extra', 'armv7', 'cp38-cp38'),
         PythonArtifact('linux_extra', 'armv7', 'cp39-cp39'),
         PythonArtifact('linux_extra', 'armv7', 'cp310-cp310', presubmit=True),
+        PythonArtifact('musllinux_1_1', 'x64', 'cp310-cp310', presubmit=True),
```
Comment (jtattermusch) [inline tools/run_tests/artifacts/artifact_targets.py:422]: given all the recent changes/speedup in the build_artifacts related jobs, I'd like to see a summary on how much extra time these artifacts add  on presubmit and on master.
I really don't want to annulate all the speedup we just did by adding 10 new artifacts.

If the numbers don't look good, we'll need to come up with more ideas how to speed things up on master and presubmit.

File: tools/internal_ci/linux/grpc_build_artifacts.cfg [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 # Location of the continuous shell script in repository.
 build_file: "grpc/tools/internal_ci/linux/grpc_build_artifacts.sh"
-timeout_mins: 180
+timeout_mins: 240
```
Comment (jtattermusch) [inline tools/internal_ci/linux/grpc_build_artifacts.cfg:19]: let's revert this change.

Comment (lidizheng) [reply tools/internal_ci/linux/grpc_build_artifacts.cfg]: Done.

File: tools/run_tests/artifacts/artifact_targets.py (hunk 1 above)
Comment (lidizheng) [reply tools/run_tests/artifacts/artifact_targets.py:422]: In the description of https://github.com/grpc/grpc/pull/28228, I updated the latest time spend on each job. This PR should only affect the Linux artifact build job and the Python distribtest job. The time spans in parenthesis order in: before optimization, after optimization, this PR.

- grpc/core/pull_request/linux/grpc_build_artifacts (170 mins -> 68 mins -> 79 mins)
- grpc/core/pull_request/linux/grpc_distribtests_python (120 mins -> 29mins -> 54 mins)

They are not the slowest presubmit checks as of today.

**Comment by lidizheng** [comment]
@jtattermusch PTALAA.

File: tools/run_tests/artifacts/artifact_targets.py (hunk 1 above)
Comment (ctiller) [reply tools/run_tests/artifacts/artifact_targets.py:422]: I'd like us to be targeting <20 minutes across the board for pull requests.

Comment (lidizheng) [reply tools/run_tests/artifacts/artifact_targets.py:422]: Well, we need significant improvement in the infrastructure to achieve this. In past 2 weeks, Jan and I have reached for low hanging fruits we can find, and the result is reducing the build time to 1/3 or 1/2.

Feel free to share your thought on go/grpc-ci-options. We are discussing the next step of our infrastructure.

Comment (jtattermusch) [reply tools/run_tests/artifacts/artifact_targets.py:422]: I agree with Craig that we should be more aggressive in terms of reducing the duration of PR jobs. It would be nice if we could come up with more speedups and only then added more stuff that makes the builds slower.

Comment (jtattermusch) [reply tools/run_tests/artifacts/artifact_targets.py:422]: FTR, I think the grpc_build_artifacts slowdown potentially added by this PR is partially masked by the fact that the all the ruby artifacts are built serially (in a single 'task') and the long tail of the ruby artifact build makes to job finish much later than it otherwise could. I'm planning to remove that limitation in https://github.com/grpc/grpc/pull/28243, and once that PR is in, the negative effect of adding the musllinux artifacts will become more significant. I don't think we want that.

Comment (lidizheng) [reply tools/run_tests/artifacts/artifact_targets.py:422]: @jtattermusch What can I do to push this PR forward?

**Comment by lidizheng** [comment]
@jtattermusch Is there any other blocker for this PR?

**Comment by jtattermusch** [comment]
> @jtattermusch Is there any other blocker for this PR?

I think the blocker is still the same - the artifact build is still pretty slow (both on PRs and on master) and we've invested significantly into speeding it up. I don't want to annulate the recent improvements by adding yet more artifacts to build.

**Comment by rushilsrivastava** [comment]
> One other option to consider is to dramatically cut the # of python versions supported for musl? Just supporting a few recent versions could help reduce the cost. Given alpine linux has an end of life support https://endoflife.date/alpine that currently includes only up to `v3.12` which includes [python 3.8.10](https://pkgs.alpinelinux.org/packages?name=python3&branch=v3.12) it seems nice to drop support for 3.6 and 3.7.

I also agree with this, especially if this can help this get merged in. Doesn't really make sense to have support for 3.6/3.7 when its on EOL on Alpine. At minimum, 3.6 can certainly be dropped since it's EOL in general.

**Comment by ctiller** [comment]
How frequently do we think these tests will catch something on presubmit that's not caught by other things?
Another way of asking: does this need to be a presubmit, or could it run on master only?

**Comment by lidizheng** [comment]
@ctiller I would recommend to have at least one Python build on musl libc. Usually, it's the Core code that needs this check. But I removed them from the presubmit checks. Hope this PR can be part of 1.44.

**Comment by lidizheng** [comment]
@jtattermusch PTALA. We have collected the speed up for the presubmits. This PR won't affect the result, so I put the artifacts and distribtests back to presubmit. Can we check it in now?

File: tools/dockerfile/grpc_artifact_python_musllinux_1_1_x86/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+FROM quay.io/pypa/musllinux_1_1_i686:2021-11-15-a808c18
+
+###################################
+# Install Python build requirements
+RUN /opt/python/cp36-cp36m/bin/pip install --upgrade cython
+RUN /opt/python/cp37-cp37m/bin/pip install --upgrade cython
+RUN /opt/python/cp38-cp38/bin/pip install --upgrade cython
+RUN /opt/python/cp39-cp39/bin/pip install --upgrade cython
+RUN /opt/python/cp310-cp310/bin/pip install --upgrade cython
```
Comment (jtattermusch) [inline tools/dockerfile/grpc_artifact_python_musllinux_1_1_x86/Dockerfile:23]: add ccache as for the other python artifact images - so that we can gain speedup for musl artifacts as well (and make sure builds are being sped up by ccache)

File: tools/dockerfile/grpc_artifact_python_musllinux_1_1_x64/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+FROM quay.io/pypa/musllinux_1_1_x86_64:2021-11-15-a808c18
+
+###################################
+# Install Python build requirements
+RUN /opt/python/cp36-cp36m/bin/pip install --upgrade cython
+RUN /opt/python/cp37-cp37m/bin/pip install --upgrade cython
+RUN /opt/python/cp38-cp38/bin/pip install --upgrade cython
+RUN /opt/python/cp39-cp39/bin/pip install --upgrade cython
+RUN /opt/python/cp310-cp310/bin/pip install --upgrade cython
```
Comment (jtattermusch) [inline tools/dockerfile/grpc_artifact_python_musllinux_1_1_x64/Dockerfile:23]: add ccache as for the other python artifact images - so that we can gain speedup for musl artifacts as well

**Review by jtattermusch — COMMENTED** [review]
What's the slowdown for build_artifact and python_distribtest jobs on PRs and on master
after adding the musl artifacts? (after you add ccache support as pointed out in the comments below)

**Review by lidizheng — COMMENTED** [review]
With ccache enabled, the Linux artifact build time reduced from [37 mins](https://source.cloud.google.com/results/invocations/762ef960-f45d-4b60-ba7e-5db8d9205a0d/targets) to [28 mins](https://source.cloud.google.com/results/invocations/938a5f3f-1289-4adc-968c-c1bc559fd515/targets) with the additional musllinux artifacts.

@jtattermusch PTALAA.

File: tools/internal_ci/linux/grpc_build_artifacts.cfg (hunk 2 above)
Comment (jtattermusch) [reply tools/internal_ci/linux/grpc_build_artifacts.cfg:19]: Looks like this change is still there? Let's revert as it really should be unnecessary.

File: tools/run_tests/artifacts/artifact_targets.py [hunk 5]
```diff
@@ -444,6 +460,16 @@ def targets():
...
+        PythonArtifact('musllinux_1_1', 'x64', 'cp36-cp36m', presubmit=True),
+        PythonArtifact('musllinux_1_1', 'x64', 'cp37-cp37m'),
+        PythonArtifact('musllinux_1_1', 'x64', 'cp38-cp38'),
+        PythonArtifact('musllinux_1_1', 'x64', 'cp39-cp39'),
+        PythonArtifact('musllinux_1_1', 'x86', 'cp310-cp310', presubmit=True),
```
Comment (jtattermusch) [inline tools/run_tests/artifacts/artifact_targets.py:468]: qq: do we really need support for 32bit musl linux? I'd have hoped that 32bit is long dead.

**Review by jtattermusch — APPROVED** [review]
LGTM with a few very minor comments.

Thanks for you patience and for the work toward speeding up artifact builds, I think it was well worth it.
